http-client/src/main/java/io/airlift/http/client/JsonStreamingResponseHandler.java

```diff
@@ -0,0 +1,184 @@
+package io.airlift.http.client;
+
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonToken;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectReader;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ListMultimap;
+import com.google.common.primitives.Ints;
+import io.airlift.json.JsonCodec;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.UncheckedIOException;
+import java.nio.charset.StandardCharsets;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+import static io.airlift.http.client.JsonResponseHandler.validateResponse;
+import static io.airlift.http.client.ResponseHandlerUtils.propagate;
+import static java.util.Objects.requireNonNull;
+
+public class JsonStreamingResponseHandler<T>
+        implements StreamingResponseHandler<T>
+{
+    private final ObjectMapper objectMapper;
+    private final ObjectReader objectReader;
+    private final Set<Integer> successfulResponseCodes;
+
+    public static <T> JsonStreamingResponseHandler<T> createJsonStreamingResponse(JsonCodec<T> codec)
+    {
+        return new JsonStreamingResponseHandler<>(codec, 200, 201, 202, 203, 204, 205, 206);
+    }
+
+    public static <T> JsonStreamingResponseHandler<T> createJsonStreamingResponse(JsonCodec<T> codec, int firstSuccessfulResponseCode, int... otherSuccessfulResponseCodes)
+    {
+        return new JsonStreamingResponseHandler<>(codec, firstSuccessfulResponseCode, otherSuccessfulResponseCodes);
+    }
+
+    private JsonStreamingResponseHandler(JsonCodec<T> codec, int firstSuccessfulResponseCode, int... otherSuccessfulResponseCodes)
+    {
+        objectMapper = codec.mapper();
+        JavaType javaType = objectMapper.getTypeFactory().constructType(codec.getType());
+        objectReader = objectMapper.readerFor(javaType);
+
+        successfulResponseCodes = ImmutableSet.<Integer>builder().add(firstSuccessfulResponseCode).addAll(Ints.asList(otherSuccessfulResponseCodes)).build();
+    }
+
+    @Override
+    public StreamingResponseIterator<T> handle(Request request, StreamingResponse response)
+    {
+        try {
+            validateResponse(request, response, successfulResponseCodes);
+        }
+        catch (UnexpectedResponseException e) {
+            throw cleanup(request, null, response, e);
+        }
+
+        JsonParser parser;
+        try {
+            parser = objectMapper.createParser(new InputStreamReader(response.getInputStream(), StandardCharsets.UTF_8));
```

Comment:
shouldn't
```
new InputStreamReader(response.getInputStream(), StandardCharsets.UTF_8)
```
be moved to previous line using try-with-resources?

Reply:
No - the need here is that the parser is not closed. The caller has the responsibility of closing.